### PR TITLE
Unity master new unitychanges boehm cherrypicks

### PR DESCRIPTION
**Pull request by UnityAlex**

### CODE REVIEW

File: mono/metadata/domain.c [hunk 1]
```diff
@@ -984,6 +984,7 @@ mono_domain_set_internal_with_options (MonoDomain *domain, gboolean migrate_exce
 
 	SET_APPDOMAIN (domain);
 	SET_APPCONTEXT (domain->default_context);
+    mono_gc_wbarrier_generic_nostore (&domain->default_context);
```
Comment (UnityAlex) [inline mono/metadata/domain.c]: I'll fix spacing

Comment (joncham) [inline mono/metadata/domain.c]: Whitespace

File: mono/metadata/gc-internals.h [hunk 2]
```diff
@@ -62,7 +62,7 @@
 } while (0)
 
 /* useful until we keep track of gc-references in corlib etc. */
-#define IS_GC_REFERENCE(class,t) (mono_gc_is_moving () ? FALSE : ((t)->type == MONO_TYPE_U && (class)->image == mono_defaults.corlib))
+#define IS_GC_REFERENCE(class,t) (mono_gc_needs_write_barriers() ? FALSE : ((t)->type == MONO_TYPE_U && (class)->image == mono_defaults.corlib))
```
Comment (joncham) [inline mono/metadata/gc-internals.h:65]: I think we can remove this check, but no need in this PR.

File: mono/metadata/domain.c (hunk 1 above)
Comment (UnityAlex) [reply mono/metadata/domain.c]: No fair I commented first that I was going to fix it!

**Review by joncham — APPROVED** [review]

**Pull request merged**